src/new_hope/week3.clj

```diff
@@ -0,0 +1,137 @@
+(ns new-hope.week3
+  (:require [midje.sweet :refer :all]
+            [criterium.core :as crit]))
+
+;; DAY 1
+
+;; To Tree, or not to Tree
+(defn binary-tree? [s]
+  (and
+    (= (count s) 3)
+    (every?
+      #(or
+        (nil? %)
+        (and
+          (sequential? %)
+          (binary-tree? %)))
+      (rest s))))
+
+(facts "about binary-tree?"
+       (binary-tree? '(:a (:b nil nil) nil)) => true
+       (binary-tree? '(:a (:b nil nil))) => false
+       (binary-tree? [1 nil [2 [3 nil nil] [4 nil nil]]]) => true
+       (binary-tree? [1 [2 nil nil] [3 nil nil] [4 nil nil]]) => false
+       (binary-tree? [1 [2 [3 [4 nil nil] nil] nil] nil]) => true
+       (binary-tree? [1 [2 [3 [4 false nil] nil] nil] nil]) => false
+       (binary-tree? '(:a nil ())) => false)
+
+;; Beauty is Symmetry
+(defn sym?
+  ([t]
+   (let [l (second t)
+         r (last t)]
+     (sym? l r)))
+  ([tl tr]
+   (let [[lv ll lr :as tl] tl
+         [rv rl rr :as tr] tr]
+     (or
+       (and
+         (nil? tl)
+         (nil? tr))
+       (and
+         (sequential? tl)
+         (sequential? tr)
+         (= lv rv)
+         (sym? ll rr)
+         (sym? lr rl))))))
+
+(facts "about sym?"
+       (sym? '(:a (:b nil nil) (:b nil nil))) => true
+       (sym? '(:a (:b nil nil) nil)) => false
+       (sym? '(:a (:b nil nil) (:c nil nil))) => false
+       (sym? [1 [2 nil [3 [4 [5 nil nil] [6 nil nil]] nil]]
+            [2 [3 nil [4 [6 nil nil] [5 nil nil]]] nil]]) => true
+       (sym? [1 [2 nil [3 [4 [5 nil nil] [6 nil nil]] nil]]
+            [2 [3 nil [4 [5 nil nil] [6 nil nil]]] nil]]) => false
+       (sym? [1 [2 nil [3 [4 [5 nil nil] [6 nil nil]] nil]]
+            [2 [3 nil [4 [6 nil nil] nil]] nil]]) => false
+       (sym? '(:a (nil nil nil) nil)) => false)
+
+;; DAY 2
+
+;; Flipping out
+(defn flip [func] (fn [& args] (apply func (reverse args))))
+
+(facts "about flip"
+       ((flip nth) 2 [1 2 3 4 5]) => 3
+       ((flip >) 7 8) => true
+       ((flip quot) 2 8) => 4
+       ((flip take) [1 2 3 4 5] 3) => [1 2 3])
+
+;; Rotate a sequence
+(defn rotate [n coll]
+  (if (empty? coll)
+    coll
+    (let [total (count coll)
+          n (if (< n 0)
```

Comment:
Try using `mod` instead or `rem`, it handles negative numbers in a way that we need here.